### PR TITLE
Override InputStream#read(byte[], int, int) for performance

### PR DESCRIPTION
**Pull request by schlosna**
The following backtrace shows previously doing single byte read() which is inefficient, so lets support bulk reads.

```
com.sun.crypto.provider.CipherCore.update(byte[], int, int)
com.sun.crypto.provider.AESCipher.engineUpdate(byte[], int, int)
javax.crypto.Cipher.update(byte[], int, int)
javax.crypto.CipherInputStream.getMoreData()
javax.crypto.CipherInputStream.read()
com.palantir.hadoop.cipher.FsCipherInputStream.read()
java.io.InputStream.read(byte[], int, int)
```

### CODE REVIEW

**Comment by schlosna** [comment]
@ellisjoe @pwoody @robert3005 @markelliot for SA

**Comment by uschi2000** [comment]
Yeah, I had suggested that a couple of months back, but I guess nobody every verified performance benefits. Do you have numbers or evidence, @schlosna ?

File: hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java
Comment (uschi2000) [inline hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java:113]: remove the NOTE above?

**Comment by schlosna** [comment]
@uschi2000 I could probably get numbers, but don't really have cycles to do so right now.

The JMH gradle plugin doesn't play well with nebula so adding microbenchmark in this project would be a pain, but I can try a separate project with this as a dependency if I get time.

File: hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java
Comment (uschi2000) [inline hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java:124]: should suffice to close the decryptedStream since it'll close the wrapped delegate

File: hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java
Comment (schlosna) [reply hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java:113]: done

File: hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java
Comment (schlosna) [reply hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java:124]: We could split this out to a separate PR, but it seems incorrect that we're not closing the CipherInputStream which finalizes the Cipher.

**Comment by schlosna** [comment]
Did some quick non-JMH tests and it shows a ~30% speed up for files of size 128MB (typical Hadoop block size).

File: hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java
Comment (uschi2000) [reply hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java:124]: ok, so let's close both here. i believe that `delegate.close()` is redundant.

**Comment by uschi2000** [comment]
👍

**Review by uschi2000 — APPROVED** [review]

**Comment by schlosna** [comment]
@uschi2000 I don't have write access to this to merge, so can you or another please merge at will and ideally tag a release to pickup downstream?

**Comment by uschi2000** [comment]
Defer to Joe.

File: hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java
Comment (ellisjoe) [reply hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java:124]: just close the decryptedStream. it'll take care of figuring out whether or not to close the delegate.

File: hadoop-crypto/src/test/java/com/palantir/hadoop/EncryptedFileSystemTest.java
Comment (ellisjoe) [inline hadoop-crypto/src/test/java/com/palantir/hadoop/EncryptedFileSystemTest.java:108]: remove this once you're closing only the decryptedStream

**Comment by ellisjoe** [comment]
need to test the new method

**Comment by schlosna** [comment]
EncryptedFileSystemTest implicitly tests this as you can tell from coverage report, but can add some unit tests verifying individual behavior and and finalization of cipher.

File: hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java
Comment (schlosna) [reply hadoop-crypto/src/main/java/com/palantir/hadoop/cipher/FsCipherInputStream.java:124]: Closing both is safe, but can adjust next week

**Pull request merged**
